macrosynergy/download/dataquery.py

```diff
@@ -285,540 +317,629 @@ def _get_token(self) -> str:
-                        params=params,
-                        delay=delay,
-                        count=count,
+                    result = self._request_thread(
+                        url=self.access_method.base_url + endpoint,
+                        params=curr_params,
+                        proxy=self.proxy,
+                        tracking_id=f"request_{ib}",
                     )
-                    # NOTE: now the new error tickers are the only error tickers,
-                    # but error messages and final_output are appended
-                    error_tickers = rec_error_tickers
-                    error_messages.extend(rec_error_messages)
-                    final_output.extend(rec_final_output)
-                    if not error_tickers:
-                        recursive_call = False
-                    elif count > 5:
-                        recursive_call = False
-                        logger.warning(
-                            f"Error requesting tickers: {', '.join(error_tickers)}. No longer retrying."
+                    download_outputs.append(result)
+                except Exception as exc:
+                    if isinstance(exc, (KeyboardInterrupt, AuthenticationError)):
+                        raise exc
+                    else:
+                        failed_batches.append(expr_batch)
+                        self.msg_errors.append(
+                            f"Batch {ib} failed with exception: {exc}"
                         )
+                        if self.debug:
+                            raise exc
+                        else:
+                            continue
 
-                except TypeError:
-                    continue
-
-        return final_output, error_tickers, error_messages
-
-    def get_ts_expression(
-        self, expressions, original_metrics, suppress_warning, show_progress, **kwargs
-    ):
-        """
-        Main driver function. Receives the Tickers and returns the respective dataframe.
-
-        :param <List[str]> expressions: categories & respective cross-sections requested.
-        :param <List[str]> original_metrics: List of required metrics: the returned
-            DataFrame will reflect the order of the received List.
-        :param <bool> suppress_warning: required for debugging.
-        :param <dict> kwargs: dictionary of additional arguments.
-        :param <bool> show_progress: used to show progress bar.
-        :return: <pd.DataFrame> df: ['cid', 'xcat', 'real_date'] + [original_metrics].
-        """
-        if self.heartbeat:
-            logger.info("Checking connection using heartbeat")
-            clause, results = self.check_connection()
-        else:
-            clause, results = True, None
+        if len(failed_batches) > 0:
+            flat_failed_batches: List[str] = list(
+                itertools.chain.from_iterable(failed_batches)
+            )
+            self.download_data(
```

Comment:
How do we keep track of the download_outputs? And merge them together. They are not part of they state till they are being unpackage below a single time... Wouldn't we lose the retryt?

Reply:
That is a mistake :facepalm:. Will be fixed. Apologies.